### ISSUE TITLE
hyperopt not working anymore

### ISSUE THREAD

**Issue by tomismrkolj** [werkkrew/freqtrade-strategies#3]
Yesterday I have update solipisis from github and also updated custom_indicators.py.
Now I cannot hyperopt solipsis anymore. I get "CustomStoploss function did not return valid stoploss".
Using the same hyperopt command that I used 3 times in a row and worked before yesterday's update.

**Comment by werkkrew** [comment]
I did make a change in the stoploss function, but I am unable to reproduce the error.

Can you give me more details?

**Comment by tomismrkolj** [comment]
Here is some more: 
![image](https://user-images.githubusercontent.com/15356681/113340595-62c73a80-932c-11eb-8d8f-0f81816090ad.png)

**Comment by werkkrew** [comment]
Sorry dude, I have no idea what to do with this.

I am running a hyperopt as we speak with the current code as well as 3 live instances of the bot.

Without knowing what your config looks like, what the command you are using is, etc. I don't know what you expect me to do.

**Comment by tomismrkolj** [comment]
Have redownloaded strategy. No success-
here is command line:
 docker-compose run --rm freqtrade hyperopt --hyperopt Solipsis3Hyp --hyperopt-loss SharpeHyperOptLoss --strategy Solipsis3opt -e 1000 --spaces buy --timerange 20210115-20210327 --timeframe 5m

and config:

    "max_open_trades": 5,
    "stake_currency": "USDT",
    "stake_amount": "unlimited",
    "tradable_balance_ratio": 0.99,
    "fiat_display_currency": "USD",
    "timeframe": "5m",
    "dry_run": true,
    "dry_run_wallet": 400,
    "cancel_open_orders_on_exit": false,
    "unfilledtimeout": {
        "buy": 10,
        "sell": 30
    },
    "bid_strategy": {
        "price_side": "bid",
        "ask_last_balance": 0.0,
        "use_order_book": false,
        "order_book_top": 1,
        "check_depth_of_market": {
            "enabled": false,
            "bids_to_ask_delta": 1
        }
    },
    "ask_strategy": {
        "price_side": "ask",
        "use_order_book": false,
        "order_book_min": 1,
        "order_book_max": 1,
        "use_sell_signal": false,
        "sell_profit_only": false,
        "ignore_roi_if_buy_signal": false
    },
    "exchange": {
        "name": "binance",
        "key": "",
        "secret": "",
        "ccxt_config": {"enableRateLimit": true},
        "ccxt_async_config": {
            "enableRateLimit": true,
            "rateLimit": 200
        },
        "pair_whitelist": ["ADA/USDT","ETH/USDT","CHZ/USDT","XRP/USDT","LUNA/USDT","ENJ/USDT","BAT/USDT","FIL/USDT","DOT/USDT","HOT/USDT","RVN/USDT","MITH/USDT","LTC/USDT","XEM/USDT","VET/USDT","SXP/USDT","MATIC/USDT","AVAX/USDT","TRX/USDT","ONE/USDT","STMX/USDT","MANA/USDT","REEF/USDT","ATOM/USDT","ANKR/USDT","EOS/USDT","LINK/USDT","SFP/USDT","DOGE/USDT","CHR/USDT","UNI/USDT","1INCH/USDT","THETA/USDT","XLM/USDT","KAVA/USDT","COTI/USDT","SUN/USDT","ONT/USDT","TFUEL/USDT","ALGO/USDT","SC/USDT","CRV/USDT","SUSHI/USDT","CTSI/USDT","DIA/USDT","OGN/USDT","VTHO/USDT","AUDIO/USDT","BCH/USDT","ALPHA/USDT","GRT/USDT","CAKE/USDT","BTT/USDT","SAND/USDT","ZIL/USDT","TROY/USDT"],
        "pair_blacklist": [
            "BNB/BTC",
            "BNB/BUSD",
            "BNB/ETH",
            "BNB/EUR",
            "BNB/NGN",
            "BNB/PAX",
            "BNB/RUB",
            "BNB/TRY",
            "BNB/TUSD",
            "BNB/USDC",
            "BNB/USDS",
            "BNB/USDT",
            "AUD/USDT",
            "BRL/USDT", 
            "CAD/USDT",  
            "CHF/USDT", 
            "CLP/USDT", 
            "CNY/USDT", 
            "CZK/USDT", 
            "DKK/USDT", 
            "EUR/USDT", 
            "GBP/USDT", 
            "HKD/USDT", 
            "HUF/USDT", 
            "IDR/USDT", 
            "ILS/USDT", 
            "INR/USDT", 
            "JPY/USDT", 
            "KRW/USDT", 
            "MXN/USDT", 
            "MYR/USDT", 
            "NOK/USDT", 
            "NZD/USDT", 
            "PHP/USDT", 
            "PKR/USDT", 
            "PLN/USDT", 
            "RUB/USDT", 
            "SEK/USDT", 
            "SGD/USDT", 
            "THB/USDT", 
            "TRY/USDT", 
            "TWD/USDT", 
            "ZAR/USDT", 
            "USD/USDT",
            "USDC/USDT",
            "BUSD/USDT",
            "SUSD/USDT",
            "TUSD/USDT",
            "BTC/USDT",
                    ]
    },
    "pairlists": [
        {"method": "StaticPairList"},
      
        {"method": "AgeFilter", "min_days_listed": 10},
        
        {"method": "PriceFilter", "low_price_ratio": 0.01},
        {"method": "SpreadFilter", "max_spread_ratio": 0.005},
        {
            "method": "RangeStabilityFilter",
            "lookback_days": 10,
            "min_rate_of_change": 0.01,
            "refresh_period": 1440
        },
        {"method": "ShuffleFilter", "seed": 42}
        ],

**Comment by werkkrew** [comment]
What is the config inside the strategy Solipsis3opt?

**Comment by tomismrkolj** [comment]
class Solipsis3opt(IStrategy):

    # Recommended for USD/USDT/etc.
    timeframe = '5m'
    inf_timeframe = '1h'

    buy_params = {
        'base-mp': 30,
        'base-rmi-fast': 50,
        'base-rmi-slow': 30,
        'inf-guard': 'upper',
        'inf-pct-adr-bot': 0.10,
        'inf-pct-adr-top': 0.85,
        'xbtc-base-rmi': 50,
        'xbtc-inf-rmi': 20,
        'xtra-base-fiat-rmi': 45,
        'xtra-base-stake-rmi': 69,
        'xtra-inf-stake-rmi': 27
    }

    sell_params = {}

    # Custom buy/sell parameters per pair
    custom_pair_params = []

    # Recommended on 5m timeframe
    minimal_roi = {
        "0": 0.01,
        "360": 0.005,
        "720": 0
    }

    dynamic_roi = {
        'enabled': True,          # enable dynamic roi which uses trennds and indicators to dynamically manipulate the roi table
        'profit-factor': 400,     # factor for forumla of how far below peak profit to trigger sell
        'rmi-start': 30,          # starting value for rmi-slow to be considered a positive trend
        'rmi-end': 70,            # ending value
        'grow-delay': 0,          # delay on growth
        'grow-time': 720,         # finish time of growth
        'fallback': 'table',      # if no trend, do what? (table, roc, atr, roc-table, atr-table)
        'min-roc-atr': 0.0075     # minimum roi value to return in roc or atr mode
    }

    use_custom_stoploss = True
    custom_stop = {
        # Linear Decay Parameters
        'decay-time': 1080,       # minutes to reach end, I find it works well to match this to the final ROI value
        'decay-delay': 0,         # minutes to wait before decay starts
        'decay-start': -0.30,     # starting value: should be the same or smaller than initial stoploss
        'decay-end': -0.03,       # ending value
        # Profit and TA  
        'cur-min-diff': 0.03,     # diff between current and minimum profit to move stoploss up to min profit point
        'cur-threshold': -0.02,   # how far negative should current profit be before we consider moving it up based on cur/min or roc
        'roc-bail': -0.03,        # value for roc to use for dynamic bailout
        'rmi-trend': 50,          # rmi-slow value to pause stoploss decay
        'bail-how': 'immediate',  # set the stoploss to the atr offset below current price, or immediate
        # Positive Trailing
        'pos-trail': False,       # enable trailing once positive  
        'pos-threshold': 0.005,   # trail after how far positive
        'pos-trail-dist': 0.015   # how far behind to place the trail
    }

    stoploss = custom_stop['decay-start']

    # Recommended
    use_sell_signal = False
    sell_profit_only = False
    ignore_roi_if_buy_signal = False

    # Required
    startup_candle_count: int = 72
    process_only_new_candles = False

    # Strategy Specific Variable Storage
    custom_trade_info = {}
    custom_fiat = "USDT" # Only relevant if stake is BTC or ETH
    custom_current_price_cache: TTLCache = TTLCache(maxsize=100, ttl=300) # 5 minutes

**Comment by werkkrew** [comment]
I'll see if I can re-create the issue on my side but unless I can, I have no idea.

**Comment by tomismrkolj** [comment]
It is also happening during dryrun:
2021-04-01 20:10:40 - freqtrade.strategy.interface - WARNING - CustomStoploss function did not return valid stoploss


Bot runs, but these messages are spread around log. It must have something with update. Since dryrun is on raspberry pi, and hyperopt was tried on pc. i copied yesterday strategy to pi and pc, and also custom_indicators.py. config.json did not change on pi!

**Comment by werkkrew** [comment]
Well you can look at the previous commit and change it back, no?

**Comment by tomismrkolj** [comment]
There is another thing: freqtrade was updated few days ago. At least on pc i pulled new image. I am not sure about pi. I will try reverting to previous commit of strategy file.
What about custom_indicators.py? Which changes have been done?

**Comment by werkkrew** [comment]
I'm sorry if I'm being short but 100% of the questions you ask can be
answered simply by looking at the code or understanding git and looking at
the commits.  I did not sign up to support and hand-hold people through
using this strategy when I released it.

I have just updated my local copy of freqtrade with the most recent changes
and I am trying to re-create your issue.

**Comment by werkkrew** [comment]
Okay, I updated my local copy to the latest develop branch from upstream.  I am not able to recreate **part** of your issue, the confirm_trade_entry piece.  I fixed it by simply removing that part of code from my copy.

I am still unable to recreate the stoploss issue you are seeing.

**Comment by werkkrew** [comment]
I believe I have identified the issue and it has resulted in some significant changes to the custom_stoploss.  Admittedly, I had done it completely wrong due a lack of understanding of the return values Freqtrade wants.

I am testing my changes/fixes now and my next push should resolve this issue for you.

**Comment by tomismrkolj** [comment]
I did not quite understand: did or you did not have problems with confirm_trade_entry? In previous answer you wrote that you removed that piece of code from copy (of freqtrade?)

**Comment by werkkrew** [comment]
It would seem there have been some significant changes upstream that broke a couple things here.  Once I updated to the latest version from freqtrade:develop I was able to reproduce the errors you were seeing.

1) confirm_trade_entry started to produce an error, and rather than troubleshoot it, I simply removed it because it was not critical to the strategy.

2) I realized two things about the custom_stoploss - 1) my return value was completely wrong.  I was returning a value of where I want the stoploss to be relative to the purchase price, but freqtrade wants it as a value relative to the current price. 2) Something else changed upstream that was making it so that negative values being returned are no longer valid (I think).  So I had to re-work it quite a bit.  It appears to be working the way I had originally intended it to now, without errors.  However I would recommend doing some additional testing as it will not behave at all like it did before in practice.

I am doing some of my own testing now so if I encounter any additional issues I will try to fix them asap.

**Comment by tomismrkolj** [comment]
I cannot say i am glad to be the one finding this issue, but it seems this episode will improve your strategy altogether.

**Comment by werkkrew** [comment]
It might, or it might break the illusion it was good to begin with, we will see.

**Comment by werkkrew** [comment]
I have pushed the changes, let me know how it works out.

**Comment by tomismrkolj** [comment]
Ok, with new files hyperopt is working. No errors!
Now we wait for the results. It usually takes 17 hours for 1000 epochs. So tomorrow sometimes.

**Comment by werkkrew** [comment]
I made a couple additional adjustments just now, you might want to grab the latest and restart the hyperopt before it gets too far along.

I **hope** there wont be any more changes for a bit.

**Comment by tomismrkolj** [comment]
Tnx, I was at 2%. Restarted hyperopt.

**Comment by tomismrkolj** [comment]
Unfortunately preliminary results after 85 epoch don't look good.
![image](https://user-images.githubusercontent.com/15356681/113449206-9a51e780-93fd-11eb-9d5d-d1b4e21fe22b.png)

This was previous version that worked before freqtrade update and which I am still running dry with success:
![image](https://user-images.githubusercontent.com/15356681/113449303-d422ee00-93fd-11eb-9d0f-fb17f89cfdce.png)

**Comment by werkkrew** [comment]
Sorry?

Not sure what to say, the "corrected" stoploss is a lot more aggressive than it was before (because its actually doing what its supposed to do), I think.

Maybe alter the parameters there to make it decay more slowly of hit a higher minimum endpoint?

**Comment by tomismrkolj** [comment]
Looks significantly better than mine. I will wait with this parameters till 1000 epochs. But if it wont be much better I will probably have to modify customstoploss by hand.
So you suggest that i increase 'decay-time' over 1080 minutes or 'decay-end' to more negative than 3%?

**Comment by werkkrew** [comment]
> 
> 
> Looks significantly better than mine. I will wait with this parameters till 1000 epochs. But if it wont be much better I will probably have to modify customstoploss by hand.
> So you suggest that i increase 'decay-time' over 1080 minutes or 'decay-end' to more negative than 3%?

Probably.  Just guessing based on how the custom_stoploss was behaving before, the decay was often not ever catching up to the trade due to the mistake I had made - that is why previous version had such a better w/l ratio - the decay stoploss wasn't doing what I thought it was.

If you think about it, I thought I was returning an end value of -0.03 meaning, by the end of the decay the stoploss was -3% and any trade worse than that would get stop lossed out.  Instead, what was happening was the decay was going to -3% below the current profit point of the trade, and thus never actually catching it. 

This lead to very few stoplosses and allowing trades much more freedom to stay open and recover.

If my theory is correct, you can sort of emulate this behavior by giving more room for trades to stay negative without hitting stoploss (lower end value) or giving more time for it to reach the end value.

This is all assuming it is actually working how I expect it to be now, which based on my results (lower avg. profit, higher # of losses), I think it is.

**Comment by tomismrkolj** [comment]
I guess it is more feasable to increase decay-end, as decay-time is already quite long. But in order to have most optimal settings I would need to run customstoploss and dynamicROI hyperopt as well. Which as I understand can be done if i install your branch of freqtrade. I am running docker versions on PC and RPi4. Is the procedure for installing your branch the same if i modify docker-compose.yml to pull from your repo?

**Comment by werkkrew** [comment]
No, unfortunately, I haven't made a container with those changes in it, I am just running a local virtualenv installation so its easy just to pull from my repo and switch to the hyperopt branch.

I will look into what I have to do to build a container from my hyperopt branch.

**Comment by werkkrew** [comment]
Just as an FYI, I am going to close this but I wanted to let you know I am actively working on Solipsis v4, I hope to have significant improvements to the buy, custom_stoploss, and dynamic_roi in the next version.

I will also try to build a docker image for my hyperopt repo and improve my documentation a bit.

Keep an eye out.
